### ISSUE TITLE
Usage of `bail!(error)` loses the context of `error`

### ISSUE THREAD

**Issue by alexcrichton** [dtolnay/anyhow#46]
I was in a situation recently where I was doing:

```rust
match some_result {
    Ok(()) => {}
    Err(e) => {
        if !err_is_ok(e) {
            bail!(e);
        }
    }
}
```

in this case `e` was actually of type `anyhow::Error` so a `return Err(e)` would have worked just fine, but I was surprised when this `bail!()` lost the context of the error of `e`, presumably because it called `.to_string()` on the result and lost all the context information.

**Comment by dtolnay** [comment]
Sorry about the confusion! This was a bug, I've fixed it in 1.0.22. It should now correctly preserve context for anyhow::Error as the argument as well as Box\<dyn Error\> and Box\<dyn Error + Send + Sync\>.

**Comment by alexcrichton** [comment]
👍
